switools/swisignature.py

```diff
@@ -144,8 +159,121 @@ def signSwiHandler( args ):
    rootCaFile = args.CAfile
    signatureFile = args.signature
    signingKeyFile = args.key
-   signSwi( swi, signingCertFile, rootCaFile, signatureFile, signingKeyFile )
-   print( 'SWI/X file %s successfully signed and verified.' % swi )
+   with tempfile.TemporaryDirectory( prefix="swi-signature-" ) as workDir:
+      signSwiAll( workDir, swi, signingCertFile, rootCaFile, signatureFile, signingKeyFile )
+      print( 'SWI/X file %s successfully signed and verified.' % swi )
+
+def runCmd( cmd, workDir ):
+   try:
+      subprocess.check_call( cmd.split(" "), cwd=workDir )
+   except subprocess.CalledProcessError:
+      return False
+   return True
+
+def insertSignature( swi, sigFileName, workDir ):
+   cmd = "zip -q -0 -X %s %s" % ( os.path.abspath( swi ), sigFileName )
+   if not runCmd( cmd, workDir ):
+      msg = "Error: Cannot insert signature file '%s' into '%s'" % ( sigFileName, swi )
+      raise SwiSignException( SWI_SIGN_RESULT.ERROR_SIGNATURE_INSERTION_FAILED, msg )
+
+def extractSignature( swi, destFile ):
+   sigFn = signaturelib.SWI_SIG_FILE_NAME
+   destDir = os.path.dirname( destFile )
+   cmd = "unzip -o -q %s %s" % ( os.path.abspath( swi ), sigFn )
```

Comment:
Please use `zipfile`.

Reply:
zipfile.py does not respect timestamps (which is pathetic), which are crucial in this case, since they affect the image's checksum, so have to be preserved across extraction/re-insertion-after-rename. Asking the zipfile what the original timestamp was, then use os.touch (does that exist) to fix it up after the fact is just jumping over backwards for no extra benefit.

Reply:
Yes, that's too much effort to workaround the bugs in `zipfile`.  Sorry, I didn't know it had those bugs.  `unzip` is preferrable here.